rfd/0149-headless-kube.md

```diff
@@ -0,0 +1,257 @@
+		return 0, trace.Wrap(err, "memfd create")
+	}
+	// set the size of the file to the size of the byte slice
+	err = unix.Ftruncate(fd, int64(len(b)))
+	if err != nil {
+		return 0, trace.Wrap(err, "ftruncate")
+	}
+
+	// map the file into memory
+	data, err := unix.Mmap(fd, 0, len(b), unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED)
+	if err != nil {
+		return 0, trace.Wrap(err, "mmap memory")
+	}
+	copy(data, b)
+
+	// unmap the file from memory
+	err = unix.Munmap(data)
+	if err != nil {
+		return 0, trace.Wrap(err, "munmap memory")
+	}
+
+	return fd, nil
+}
+
+func reexec(data string) {
+	fd, err := memFile("proxy-kubeconfig", []byte(data))
+	if err != nil {
+		panic(err)
+	}
+
+	// filepath to our newly created in-memory file descriptor
+	fp := fmt.Sprintf("/proc/self/fd/%d", fd)
+
+	// open the file
+	f := os.NewFile(uintptr(fd), fp)
+	defer f.Close()
+
+	// prepare to re-exec shell
+	command := "/bin/bash"
+	if shell, ok := os.LookupEnv("$SHELL"); ok {
+		command = shell
+	}
+
+	cmd := exec.Command(command)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+	
+    // pass the file descriptor to the child process as an extra file
+    // descriptor. It will be available as fd 3 in /proc/self/fd/3
+    cmd.ExtraFiles = []*os.File{f}
+	
+	cmd.Env = os.Environ()
+	// Point $KUBECONFIG to the in-memory file so kubectl is ready to use immediately.
+    cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", teleport.EnvKubeConfig, "/proc/self/fd/3"))
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	if err := cmd.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+```
+
+## Diagram
+
+Headless kube proxy flow shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local as Local Machine
+    participant Headless as Remote (Headless) Machine
+    participant Teleport
+
+    Note over Headless: tsh --headless proxy kube MainKubeCluster
+
+    Headless->>Teleport: POST /webapi/login/headless
+
+    Headless-->>Local: user copies URL to local browser
+    Note over Local: proxy.example.com/headless/<request_id>
+
+    Local->>Teleport: user approves headless request
+    Teleport->>Local: MFA Challenge
+    Note over Local: user taps YubiKey<br/>to sign MFA challenge
+    Local->>Teleport: rpc UpdateHeadlessAuthenticationState<br/>with signed MFA challenge response
+
+    Teleport->>Headless: user certificates<br/>(MFA-verified, 1 minute TTL)
+    Headless->>Teleport: request certificate for accessing MainKubeCluster
+    Teleport->>Headless: MainKubeCluster certificate<br/>(max_session_ttl TTL)
+    Headless->>Headless: kube proxy process reexecs itself, starting new shell<br/>proxy still running keeping kubeconfig in memory-only file<br/>new shell has $KUBECONFIG set to the memory-only file
+    Note over Headless: kubectl get pods
+```
+## UX
+User experience while using the headless Kubernetes access will look like this:
+
+```console
+$ tsh --headless proxy kube main_kube_cluster --proxy=tele.port --user=admin
```

Comment:
I am reviewing the actual PR but I think it's easier to ask in the RFD.
What happens if multiple kube clusters are passed like `tsh --headless proxy kube kube1 kube2 --proxy=tele.port --user=admin`
 Or if no kube clusters are passed at all like `tsh --headless proxy kube --proxy=tele.port --user=admin`

Reply:
It works similarly to the regular `tsh proxy kube`. If multiple clusters are passed it will get certs for all those clusters. The case with no clusters provided makes no sense though for the headless mode, since you're supposedly using it on a machine which doesn't have your tsh credentials. Right now in PR it tries to look it up, but I'll change it to error out in that case and tell user to provide cluster(s) name 👍